File: modulo08-arquitetura-de-sistemas-com-ia/modulo-02-single-agent/agent-components-demo-java/src/main/java/com/trialforge/agentcomponents/Planejamento.java
package com.trialforge.agentcomponents;

import java.io.IOException;
import java.util.List;

/**
 * 2. PLANEJAMENTO — chain-of-thought (quase de graca) vs reflexao (chamada a mais).
 *
 * Unica secao que chama o modelo de verdade (Ollama) — por isso, diferente de
 * Memoria/Ferramentas/AcaoGate, nao tem teste automatizado equivalente no
 * original (so observacao ao vivo em demonstrarPlanejamento()).
 *
 * Porte 1:1 de agent-components-demo.js / agent_components_demo.py (secao 2).
 */
public final class Planejamento {

    private Planejamento() {
    }

    public record ResultadoPlanejamento(String texto, long duracaoMs, int chamadas) {
    }

    public static ResultadoPlanejamento chainOfThought(ChatClient client, String pergunta)
            throws IOException, InterruptedException {
        long inicio = System.currentTimeMillis();
        String resposta = client.chat(List.of(
                new Memoria.Mensagem("user", "Pense passo a passo, de forma breve, antes de responder: " + pergunta)
        ));
        return new ResultadoPlanejamento(resposta, System.currentTimeMillis() - inicio, 1);
    }

    public static ResultadoPlanejamento chainOfThoughtMaisReflexao(ChatClient client, String pergunta)
            throws IOException, InterruptedException {
        long inicio = System.currentTimeMillis();
        String primeira = client.chat(List.of(new Memoria.Mensagem("user", pergunta)));
        String critica = client.chat(List.of(
                new Memoria.Mensagem("user", pergunta),
                new Memoria.Mensagem("assistant", primeira),
                new Memoria.Mensagem("user",
                        "Releia sua resposta acima. Ela tem algum erro ou imprecisão? Responda só \"sim\" ou \"não\" e, se sim, qual.")
        ));
        return new ResultadoPlanejamento(critica, System.currentTimeMillis() - inicio, 2);
    }
}

File: modulo08-arquitetura-de-sistemas-com-ia/modulo-02-single-agent/agent-components-demo-go/internal/agente/planejamento.go
package agente

import (
	"context"
	"time"
)

// ChatClient e a abstracao do modelo de chat usado pela secao de
// Planejamento. Extraida como interface (o original em JS/Python chama o
// SDK do Ollama diretamente) para manter a chamada de rede isolada e
// substituivel — nao porque os testes precisem de um dublê: a secao de
// Planejamento nao tem teste automatizado, igual ao original (so
// observacao ao vivo).
type ChatClient interface {
	Chat(ctx context.Context, mensagens []Mensagem) (string, error)
}

// ResultadoPlanejamento contem o texto da resposta, quanto tempo levou, e
// quantas chamadas ao modelo foram feitas.
type ResultadoPlanejamento struct {
	Texto     string
	DuracaoMs int64
	Chamadas  int
}

// ChainOfThought pede pro modelo pensar passo a passo antes de responder,
// em uma unica chamada.
func ChainOfThought(ctx context.Context, client ChatClient, pergunta string) (ResultadoPlanejamento, error) {
	inicio := time.Now()
	resposta, err := client.Chat(ctx, []Mensagem{
		{Role: "user", Content: "Pense passo a passo, de forma breve, antes de responder: " + pergunta},
	})
	if err != nil {
		return ResultadoPlanejamento{}, err
	}
	return ResultadoPlanejamento{
		Texto:     resposta,
		DuracaoMs: time.Since(inicio).Milliseconds(),
		Chamadas:  1,
	}, nil
}

// ChainOfThoughtMaisReflexao faz uma segunda chamada, pedindo pro modelo
// criticar a propria resposta — chamada inteira a mais, nao so raciocinio
// mais longo na mesma chamada.
func ChainOfThoughtMaisReflexao(ctx context.Context, client ChatClient, pergunta string) (ResultadoPlanejamento, error) {
	inicio := time.Now()

	primeira, err := client.Chat(ctx, []Mensagem{{Role: "user", Content: pergunta}})
	if err != nil {
		return ResultadoPlanejamento{}, err
	}

	critica, err := client.Chat(ctx, []Mensagem{
		{Role: "user", Content: pergunta},
		{Role: "assistant", Content: primeira},
		{Role: "user", Content: "Releia sua resposta acima. Ela tem algum erro ou imprecisão? Responda só \"sim\" ou \"não\" e, se sim, qual."},
	})
	if err != nil {
		return ResultadoPlanejamento{}, err
	}

	return ResultadoPlanejamento{
		Texto:     critica,
		DuracaoMs: time.Since(inicio).Milliseconds(),
		Chamadas:  2,
	}, nil
}
